pandas/core/groupby/ops.py

```diff
@@ -574,9 +587,52 @@ def _ea_wrap_cython_operation(
             f"function is not implemented for this dtype: {values.dtype}"
         )
 
+    @final
+    def _masked_ea_wrap_cython_operation(
+        self,
+        kind: str,
+        values: BaseMaskedArray,
+        how: str,
+        axis: int,
+        min_count: int = -1,
+        **kwargs,
+    ) -> BaseMaskedArray:
+        """
+        Equivalent of `_ea_wrap_cython_operation`, but optimized for masked EA's
+        and cython algorithms which accept a mask.
+        """
+        orig_values = values
+
+        # isna just directly returns self._mask, so copy here to prevent
+        # modifying the original
+        mask = isna(values).copy()
+        arr = values._data
+
+        if is_integer_dtype(values.dtype) or is_bool_dtype(values.dtype):
```

Comment:
why is this an entirely different funtion? pls integrate to existing infra

Reply:
This is a different function because it's specific for MaskedArrays. Having this as a separate function is consistent with how it's currently implemented IMO (with a similar separate function for generic EAs). 

It could also be another `elif` check in `_ea_wrap_cython_operation`, but it's not that it would result in less code or so, and since `_ea_wrap_cython_operation` already gets quite complicated, I think this separate function is good.

Reply:
In an initial version I tried to fold this into `_ea_wrap_cython_operation`, but thought this smaller function was a cleaner solution since the conditionals here can remain much simpler. While adding a function for just supporting masked `cummax`, `cummin` seems wasteful, this infrastructure should extend to more masked groupby algos.

Reply:
see my comments. This MUST integrate with the existing infrastructure (or refactor that). Duplicating is -1

Reply:
1) The whole point of this PR is to *add* a special handling for masked arrays, to ensure to pass through the mask to the groupby cython algo. This will always add *some* code (and the code in `_masked_ea_wrap_cython_operation` is specific to masked arrays). 
2) IMO this *is* integrated with the existing infrastructure:  it integrates nicely into the existing `_cython_operation` and follows the same pattern as we already have for `_ea_wrap_cython_operation`

If you don't like how the added code is structured right now, please do a *concrete* suggestion of how you would do it differently.